_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,871 @@
+
+{: #mmfsr}
+
+### MemManage Status Register (MMFSR) - 0xE000ED28
+
+This register reports Memory Protection Unit faults. For a deep dive into the peripheral, see [this article](https://interrupt.memfault.com/blog/fix-bugs-and-secure-firmware-with-the-mpu).
+
+> NOTE: For the most part MPU faults will only trigger if the MPU has been enabled. However, there are a few memory access errors that will always result in a MemManage fault -- such as trying to execute from the `0xExxx.xxxx` address range.
+
+The layout of the register looks like this:
+
+![](img/cortex-m-fault/mmfsr.png)
+
+where,
+
+- `MMARVALID` - Indicates the address which triggered the MemManage fault is stored in the _MemManage Fault Address Register_ (**MMFAR**) - a 32 bit register located at `0xE000ED34`
+- `MLSPERR` & `MSTKERR` - Indicates that a MemManage fault occurred during lazy state preservation or exception entry. This could happen if you are using an MPU region as a stack guard in your system
+- `MUNSTKERR` - Indicates that a fault occurred while returning from an exception
+- `DACCVIOL` - Indicates that a data access triggered the MemManage fault.
+- `IACCVIOL` - Indicates that an attempt to execute an instruction triggered a MPU or Execute Never (XN) error.
+
+## HardFault Status Register (HFSR) - 0xE000ED2C
+
+This registers explains the reason a HardFault exception was triggered.
+
+![](img/cortex-m-fault/hfsr.png)
+
+There's not too much information in this register but we will go over the fields real quickly
+
+- `DEBUGEVT` - Indicates that a debug event (i.e executing a breakpoint instruction) occurred while the debug subsystem was not enabled
+- `FORCED` - This means a fault of configurable priority, the faults we just discussed [above](#cfsr) has been escalated to a hardfault. The exception handlers for configurable faults are not enabled by default so this is pretty common. If the exception handlers are enabled, a configurable fault may still escalate to a hardfault if the configurable fault occurred at a higher priority than the exception handler
+- `VECTTBL` - Indicates a fault occurred because of an issue reading from an address in the vector table. This is pretty atypical but could happen if there is a bad address in the vector table.
+
+{: #registers-prior-to-exception}
+
+## Recovering Register State
+
+To fix a fault, you will want to determine what code was running when the fault occurred. To do this we will want to recover the register state at the time of exception entry.
+
+If you the fault you are using is reproducible and you are trying to look at it with a debugger you can manually add a breakpoint for the address which handles the exception in the vector table. Typically in GDB this will look something like
+
+```
+(gdb) break HardFault_Handler
+```
+
+However, let's discuss how you can instrument your code to make this recovery easier the next time an unexpected fault hits!
```

Comment:
One thing that's missing here is the *why*. What is hard to do if I did not instrument my code? A short example would help.

Reply:
added an example and moved the c code to the "automating" section